server/src/main/java/org/elasticsearch/search/lookup/SearchLookup.java

```diff
@@ -24,22 +24,88 @@
+    /**
+     * The maximum "depth" of field dependencies. When a field's doc values
+     * depends on another field's doc values and depends on another field's
+     * doc values that depends on another field's doc values, etc, etc,
+     * it can make a very deep stack. At some point we're concerned about
+     * {@link StackOverflowError}. This limits that "depth".
+     */
+    static final int MAX_FIELD_CHAIN = 30;
 
-    final DocLookup docMap;
+    /**
+     * The chain of fields for which this lookup was created used for detecting
+     * loops in doc values calculations. It is empty for the "top level" lookup
+     * created for the entire search. When a lookup is created for a field we
+     * make sure its name isn't in the list then add it to the end. So the
+     * lookup for the a field named {@code foo} will be {@code ["foo"]}. If
+     * that field looks up the values of a field named {@code bar} then
+     * {@code bar}'s lookup will contain {@code ["foo", "bar"]}.
+     */
+    private final List<String> fieldChain;
+    private final MapperService mapperService;
+    private final BiFunction<MappedFieldType, Supplier<SearchLookup>, IndexFieldData<?>> indexFieldDataService;
+    private final DocLookup docMap;
+    private final SourceLookup sourceLookup;
+    private final FieldsLookup fieldsLookup;
 
-    final SourceLookup sourceLookup;
+    /**
+     * Create the "top level" field lookup for a search request.
+     */
+    public SearchLookup(
+        MapperService mapperService,
+        BiFunction<MappedFieldType, Supplier<SearchLookup>, IndexFieldData<?>> indexFieldDataService
+    ) {
+        this(emptyList(), mapperService, indexFieldDataService, new SourceLookup(), new FieldsLookup(mapperService));
+    }
 
-    final FieldsLookup fieldsLookup;
+    private SearchLookup(
+        List<String> fieldChain,
+        MapperService mapperService,
+        BiFunction<MappedFieldType, Supplier<SearchLookup>, IndexFieldData<?>> indexFieldDataService,
+        SourceLookup sourceLookup,
+        FieldsLookup fieldsLookup
+    ) {
+        this.fieldChain = fieldChain;
+        this.mapperService = mapperService;;
+        this.indexFieldDataService = indexFieldDataService;
+        this.docMap = new DocLookup(mapperService, ft -> this.indexFieldDataService.apply(ft, () -> forField(ft.name())));
+        this.sourceLookup = sourceLookup;
+        this.fieldsLookup = fieldsLookup;
+    }
 
-    public SearchLookup(MapperService mapperService, Function<MappedFieldType, IndexFieldData<?>> fieldDataLookup) {
-        docMap = new DocLookup(mapperService, fieldDataLookup);
-        sourceLookup = new SourceLookup();
-        fieldsLookup = new FieldsLookup(mapperService);
+    /**
+     * Build a new lookup for a field that needs the lookup to create doc values.
+     *
+     * @throws IllegalArgumentException if it detects a loop in the fields required to build doc values
+     */
+    public SearchLookup forField(String name) {
```

Comment:
This goes along my other comments: I wonder if it is necessary to create a new SearchLookup for every field. Maybe we could add references tracking at a lower level so that callers don't need to be adapted, especially given that it's applied only to DocLookup?

Reply:
I'll have a look at pushing the tracking into doc lookup. I'm not sure there is a way to avoid adapting the callers though. I'll check!

Reply:
I looked at pushing it down and I don't really like how it turned out, mostly because I need a reference to the `SearchLookup` in the `DocLookup` then which feels pretty circular.

Reply:
Can I see it? curious how it turns out.